Add event tracing and ETDumps to executor_runner

benkli01 opened the pull request:
- Enabled via EXECUTORCH_ENABLE_EVENT_TRACER
- Add flag 'etdump_path' to specify the file path for the ETDump file
- Add flag 'num_executions' for number of iterations to run
- Create and pass event tracer 'ETDumpGen'
- Save ETDump to disk
- Update docs to reflect the changes

Re-upload of https://github.com/pytorch/executorch/pull/4502 to discuss with @GregoryComer.

benkli01 committed:
Add event tracing and ETDumps to executor_runner

- Enabled via EXECUTORCH_ENABLE_EVENT_TRACER
- Add flag 'etdump_path' to specify the file path for the ETDump file
- Add flag 'num_executions' for number of iterations to run
- Create and pass event tracer 'ETDumpGen'
- Save ETDump to disk
- Update docs to reflect the changes

Signed-off-by: Benjamin Klimczak <benjamin.klimczak@arm.com>
Change-Id: I7e8e8b7f21453bb8d88fa2b9c2ef66c532f3ea46

benkli01 commented:
@pytorchbot label 'partner: arm'

benkli01 commented:
@pytorchbot label ciflow/trunk

benkli01 commented:
Hi @GregoryComer. Would it be possible to run the CI on your side to see if the [issue](https://github.com/pytorch/executorch/actions/runs/10398365231/job/28795542762) from the [previous PR](https://github.com/pytorch/executorch/pull/4502) is still occurring? I'm having a hard time understanding where this comes from.

benkli01 commented:
Hi @dbort . Sorry for dragging you into this, but I saw your comment on `EXECUTORCH_SEPARATE_FLATCC_HOST_PROJECT` in the code, so I thought you might be able to help with resolving the [failing test here](https://github.com/pytorch/executorch/actions/runs/10991797622/job/30514879193?pr=5027). Any idea how to fix this?

digantdesai commented:
I don't see a CI failure anymore

benkli01 commented:
> I don't see a CI failure anymore

@digantdesai  To me `pull / test-llama-runner-qnn-linux (fp32, cmake, qnn) / linux-job (pull_request)` is showing up as failing after my latest update. The CI run for the previous version you imported did not finish for me, i.e. I could not see any results, but it did not seem to have this test included anyway.

digantdesai commented:
Yeah I see, from the main CMakeList, and qnn does have `EXECUTORCH_ENABLE_EVENT_TRACER=ON`

digantdesai commented:
Any update on this?

benkli01 commented:
Hi @digantdesai, I'm still hoping for some pointer from @dbort or you as I'm struggling to reproduce it locally and can't really make sense of the error.

benkli01 committed:
Merge branch 'pytorch:main' into add-profiling-to-xnn-executor-runner-2

benkli01 committed:
Merge branch 'pytorch:main' into add-profiling-to-xnn-executor-runner-2

benkli01 committed:
Merge branch 'pytorch:main' into add-profiling-to-xnn-executor-runner-2

benkli01 committed:
Merge branch 'pytorch:main' into add-profiling-to-xnn-executor-runner-2

benkli01 committed:
Merge branch 'pytorch:main' into add-profiling-to-xnn-executor-runner-2

cccclai commented:
The error shows up when running this script `https://github.com/pytorch/executorch/blob/main/backends/qualcomm/scripts/build.sh` based on the log. 

If you have a linux machine, can you follow https://pytorch.org/executorch/stable/build-run-qualcomm-ai-engine-direct-backend.html and see if the script fails?

benkli01 commented:
@cccclai I finally managed to reproduce the issue by running script `backends/qualcomm/scripts/build.sh` with the parameters from the CI script [here](https://github.com/pytorch/executorch/blob/main/.ci/scripts/build-qnn-sdk.sh#L17). Interestingly, the issue seems to be caused by `--job_number 2` (not my first guess). If I remove the parameter entirely, defaulting to `--job_number 16`, the issue disappears (not sure this is an acceptable solution and/or would work in the CI). I'm guessing that this is related to the TODO [here](https://github.com/pytorch/executorch/blob/2d499b3d0cf0b085e373404e5ff421a73a4a22b4/setup.py#L545). Any input on how to proceed would be much appreciated.

benkli01 committed:
Merge branch 'pytorch:main' into add-profiling-to-xnn-executor-runner-2

cccclai commented:
> @cccclai I finally managed to reproduce the issue by running script `backends/qualcomm/scripts/build.sh` with the parameters from the CI script [here](https://github.com/pytorch/executorch/blob/main/.ci/scripts/build-qnn-sdk.sh#L17). Interestingly, the issue seems to be caused by `--job_number 2` (not my first guess). If I remove the parameter entirely, defaulting to `--job_number 16`, the issue disappears (not sure this is an acceptable solution and/or would work in the CI). I'm guessing that this is related to the TODO [here](https://github.com/pytorch/executorch/blob/2d499b3d0cf0b085e373404e5ff421a73a4a22b4/setup.py#L545). Any input on how to proceed would be much appreciated.

Ah I remember that. @dbort and @Olivia-liu, any thought on this?

dbort commented on examples/portable/executor_runner/executor_runner.cpp:
```diff
@@ -151,8 +164,20 @@ int main(int argc, char** argv) {
-
-  Result<Method> method = program->load_method(method_name, &memory_manager);
+  EventTracer* event_tracer_ptr = nullptr;
+#ifdef ET_EVENT_TRACER_ENABLED
```
This function is already so long and complex, I'd like to factor out these ifdefs if possible.

You could create a class to encapsulate the event tracing, like

```
class TraceManager {
 public:
  TraceManager();
  EventTracer* get_event_tracer();
  Error write_etdump_to_file(const char* filename);
};
```

If tracing is enabled, the ctor could create the ETDump (a field), get_event_tracer can return a pointer to it, and write_to_file can open the file and write the contents. If tracing is disabled, the class is basically empty, returning a null tracer and just returning Error::NotSupported when asked to write.

Then main() can say
```
TraceManager tracer;
program->load_method(..., tracer->get_event_tracer());
...
if (tracer->get_event_tracer() != nullptr) {
  status = tracer->write_etdump_to_file(FLAGS_etdump_path.c_str());
  ET_CHECK_MSG(status == Error::Ok, ...);
}
```

benkli01 replied:
Good idea. Let me know if the implementation looks ok.

dbort commented on examples/portable/executor_runner/executor_runner.cpp:
```diff
@@ -25,21 +26,33 @@
+DEFINE_string(
+    etdump_path,
+    "model.etdump",
+    "If ETDump generation is enabled an ETDump will be written out to this path.");
```
Since this option (and thus its help string) will only be visible to users when tracing is enabled, it doesn't need to be qualified.
```suggestion
    "Write ETDump data to this path.");
```

benkli01 replied:
Done.

dbort reviewed (changes requested):
Sorry that I missed your mentions of me for so long! Thanks @cccclai for pointing me to this.

Just to double check, is this the error you're seeing?

https://github.com/pytorch/executorch/actions/runs/12161392244/job/33915905755#step:14:1039
```
gmake[2]: *** No rule to make target '../third-party/flatcc/lib/libflatccrt.a', needed by 'executor_runner'.  Stop.
```

I think I remember @Olivia-liu looking into this, and worked around it by building in release mode. (hence https://github.com/pytorch/executorch/blob/8861b9acae528f763aa54ec40677971d2265a46b/examples/devtools/build_example_runner.sh#L44) Which is still something we need to figure out. Olivia, do you know if there was a github issue tracking this problem?

dbort commented:
Also for what it's worth, I'm trying to merge https://github.com/dvidelabs/flatcc/pull/306 into upstream flatcc to let us remove `-DEXECUTORCH_SEPARATE_FLATCC_HOST_PROJECT` and similar hacks

benkli01 commented:
> ```
> gmake[2]: *** No rule to make target '../third-party/flatcc/lib/libflatccrt.a', needed by 'executor_runner'.  Stop.
> ```
Thanks @dbort , this is the error exactly.

I'm not sure about the workaround to use release mode. The command used in the ci script is already using `--release`. As mentioned above, removing `--num_jobs 2` seems to work locally, but it's a strange workaround and might not work in CI.

dbort commented:
> I'm not sure about the workaround to use release mode. The command used in the ci script is already using --release.

Ok, thanks for looking into that.

> As mentioned above, removing --num_jobs 2 seems to work locally, but it's a strange workaround and might not work in CI.

That means that there's some kind of race condition.

Based on your PR, it looks like executor_runner has a proper dependency on libflatccrt; otherwise I would have expected a use-without-dependency situation.

...except maybe it doesn't. This PR adds a dep on "${FLATCCRT_LIB}" from the top-level CMakeLists.txt, but when I look for a place that sets FLATCCRT_LIB I only see it in the cmake config file at https://github.com/pytorch/executorch/blob/538bfaf3d5a25baf2d094c178b0fe9e708ed16f2/build/executorch-config.cmake#L51-L55

afaik, that config isn't included in the top-level cmake system. That file is used to point to an already-built version of the core ET libs from external projects, like https://github.com/pytorch/executorch/blob/538bfaf3d5a25baf2d094c178b0fe9e708ed16f2/examples/devtools/CMakeLists.txt#L39

If `FLATCCRT_LIB` is empty in this PR at https://github.com/pytorch/executorch/pull/5027/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR817-R819
```
list(APPEND _executor_runner_libs etdump ${FLATCCRT_LIB})
```
then executor_runner wouldn't properly depend on libflatccrt.a. A parallel build could cause that lib to be coincidentally built earlier with -j16, "fixing" the problem, while -j2 would be less likely to do so.

Could you try printing the value of FLATCCRT_LIB from the top-level CMakeLists.txt to see if it's empty?

dbort commented:
Though theoretically executor_runner shouldn't even need to know about libflatccrt: it should inherit the dep from the PUBLIC section of

https://github.com/pytorch/executorch/blob/538bfaf3d5a25baf2d094c178b0fe9e708ed16f2/devtools/CMakeLists.txt#L179-L183

benkli01 committed:
Merge branch 'pytorch:main' into add-profiling-to-xnn-executor-runner-2

benkli01 committed:
Merge branch 'pytorch:main' into add-profiling-to-xnn-executor-runner-2

benkli01 committed:
Fix comments during code review

- Raise a CMake error if event tracing is enabled without the devtools
- Re-factoring of the changes in the portable executor_runner
- Minor fix in docs

Change-Id: Ia50fef8172f678f9cbe2b33e2178780ff983f335
Signed-off-by: Benjamin Klimczak <benjamin.klimczak@arm.com>

benkli01 committed:
Merge branch 'pytorch:main' into add-profiling-to-xnn-executor-runner-2

benkli01 commented on examples/portable/executor_runner/executor_runner.cpp:
```diff
@@ -151,8 +164,20 @@ int main(int argc, char** argv) {
+  ET_CHECK_MSG(
+      etdump_file,
+      "Failed to open ETDump file at %s.",
+      FLAGS_etdump_path.c_str());
```
Agreed. Fixed now.

benkli01 commented on examples/portable/executor_runner/executor_runner.cpp:
```diff
@@ -171,24 +196,39 @@ int main(int argc, char** argv) {
+        method_name,
+        (uint32_t)status);
+  }
+  ET_LOG(Info, "Model executed successfully %i time(s).", FLAGS_num_executions);
```
Done.

benkli01 commented on docs/source/tutorial-xnnpack-delegate-lowering.md:
```diff
@@ -177,3 +177,6 @@ Now you should be able to find the executable built at `./cmake-out/backends/xnn
 You can build the XNNPACK backend [CMake target](https://github.com/pytorch/executorch/blob/main/backends/xnnpack/CMakeLists.txt#L83), and link it with your application binary such as an Android or iOS application. For more information on this you may take a look at this [resource](demo-apps-android.md) next.
+
+## Profiling
+To enable profiling in the `xnn_executor_runner` pass the flags `-DEXECUTORCH_ENABLE_EVENT_TRACER=ON` and `-DEXECUTORCH_BUILD_SDK=ON` to the build command (add `-DENABLE_XNNPACK_PROFILING=ON` for additional details). This will enable ETDump generation when running the inference and enables command line flags for profiling (see `xnn_executor_runner --help` for details).
```
Fixed.

benkli01 commented on CMakeLists.txt:
```diff
@@ -813,6 +814,10 @@ if(EXECUTORCH_BUILD_EXECUTOR_RUNNER)
   endif()
 
+  if(EXECUTORCH_ENABLE_EVENT_TRACER)
+    list(APPEND _executor_runner_libs etdump ${FLATCCRT_LIB})
```
Added a check for this.

benkli01 commented on backends/xnnpack/CMakeLists.txt:
```diff
@@ -128,8 +129,13 @@ if(NOT CMAKE_TOOLCHAIN_FILE MATCHES ".*(iOS|ios\.toolchain)\.cmake$")
   add_executable(xnn_executor_runner ${_xnn_executor_runner__srcs})
+
+  if(EXECUTORCH_ENABLE_EVENT_TRACER)
+    list(APPEND xnn_executor_runner_libs etdump)
```
Added a check for this.

benkli01 reviewed (commented):
Thanks for the review! All issues be fixed now.

benkli01 committed:
Merge branch 'pytorch:main' into add-profiling-to-xnn-executor-runner-2

benkli01 committed:
Small fix for use of flag `FLAGS_etdump`

Change-Id: I0ebb22636cdd64aea24bcee51cba05496ed78b1f

benkli01 committed:
Another fix for case ET_EVENT_TRACER_ENABLED=OFF

Change-Id: I7d72e4d8f46ec727a60c9553851d5b71da8e91d4
Signed-off-by: Benjamin Klimczak <benjamin.klimczak@arm.com>

benkli01 committed:
Merge branch 'pytorch:main' into add-profiling-to-xnn-executor-runner-2

benkli01 committed:
Merge branch 'pytorch:main' into add-profiling-to-xnn-executor-runner-2

dbort commented on CMakeLists.txt:
```diff
@@ -822,6 +823,14 @@ if(EXECUTORCH_BUILD_EXECUTOR_RUNNER)
 
+  if(EXECUTORCH_ENABLE_EVENT_TRACER)
+    if(EXECUTORCH_BUILD_DEVTOOLS)
+      list(APPEND _executor_runner_libs etdump ${FLATCCRT_LIB})
```
I'm still not sure about this use of FLATCCRT_LIB. flatcc defines a single target named `flatccrt`: https://github.com/dvidelabs/flatcc/blob/896db54787e8b730a6be482c69324751f3f5f117/src/runtime/CMakeLists.txt#L5, so it seems like this should be `list(APPEND _executor_runner_libs etdump flatccrt)`.

executorch's cmake build only sets FLATCCRT_LIB in https://github.com/pytorch/executorch/blob/9c043290ad3944268290e015c3063bc411e6ef6b/build/executorch-config.cmake#L52. This might be set during a build that uses `find_package(executorch CONFIG)`, but it doesn't seem to be set during a normal build.

I tried adding `message(STATUS ">>>> flatccrt_lib ${FLATCCRT_LIB}")` to the bottom of this top-level CMakeLists.txt file, and saw:
```
(rm -rf cmake-out \
    && mkdir cmake-out \
    && cd cmake-out \
    && cmake -DEXECUTORCH_ENABLE_EVENT_TRACER=ON -DEXECUTORCH_BUILD_DEVTOOLS=ON ..)

...

--   EXECUTORCH_BUILD_CPUINFO               : ON
-- >>>> flatccrt_lib
-- Configuring done (13.2s)
```
i.e., FLATCCRT_LIB is blank.

benkli01 replied:
Sorry, I had reverted things to `flatccrt` locally at some point but did not push it upstream. Fixed now.

dbort reviewed (changes requested):
Thanks for refactoring executor_runner, it looks great. Remaining issue is the flatccrt dep

benkli01 committed:
Revert use of FLATCCRT_LIB to flatccrt

Signed-off-by: Benjamin Klimczak <benjamin.klimczak@arm.com>
Change-Id: I5e7d8ef5d66bc3d5de36ea451b31fb3bdcd42d09

benkli01 committed:
Merge branch 'pytorch:main' into add-profiling-to-xnn-executor-runner-2

benkli01 committed:
Merge branch 'pytorch:main' into add-profiling-to-xnn-executor-runner-2

dbort commented:
Thanks for updating the dependency, and again I apologize for how long this has taken me to review. I'm running internal tests now and should be able to merge this soon.

benkli01 commented:
Thanks @dbort ! There is a linker error remaining in the qnn tests that I could not reproduce locally and I don't understand where it is coming from. Maybe you have an idea...

dbort reviewed (changes requested):
You're right, I see the failure
https://github.com/pytorch/executorch/actions/runs/12866353984/job/35868756676?pr=5027#step:14:34489
```
/usr/bin/ld: CMakeFiles/executor_runner.dir/examples/portable/executor_runner/executor_runner.cpp.o: in function `main':
executor_runner.cpp:(.text.main+0x5f9): undefined reference to `executorch::etdump::ETDumpGen::ETDumpGen(executorch::runtime::Span<unsigned char>)'
/usr/bin/ld: executor_runner.cpp:(.text.main+0x95d): undefined reference to `executorch::etdump::ETDumpGen::get_etdump_data()'
```

The [build output](https://github.com/pytorch/executorch/actions/runs/12866353984/job/35868756676?pr=5027#step:14:33966) says "EXECUTORCH_BUILD_DEVTOOLS : OFF", but the linker error implies that `ET_EVENT_TRACER_ENABLED` is defined.

It does seem like the block you added to the top-level CMakeLists.txt should have triggered a SEND_ERROR in this case.

For debugging, it would help to print `EXECUTORCH_ENABLE_EVENT_TRACER` in [Utils.cmake](https://github.com/benkli01/executorch/blob/add-profiling-to-xnn-executor-runner-2/build/Utils.cmake).

And to reproduce, you might be able to use the `cmake` command from the log:
https://github.com/pytorch/executorch/actions/runs/12866353984/job/35868756676?pr=5027#step:14:33817
```
cmake -DCMAKE_INSTALL_PREFIX=cmake-out -DCMAKE_BUILD_TYPE=Release -DEXECUTORCH_BUILD_EXTENSION_DATA_LOADER=ON -DEXECUTORCH_BUILD_EXTENSION_MODULE=ON -DEXECUTORCH_BUILD_EXTENSION_TENSOR=ON -DEXECUTORCH_BUILD_KERNELS_CUSTOM=OFF -DEXECUTORCH_BUILD_KERNELS_OPTIMIZED=ON -DEXECUTORCH_BUILD_KERNELS_QUANTIZED=ON -DEXECUTORCH_BUILD_XNNPACK=OFF -DEXECUTORCH_BUILD_MPS=OFF -DEXECUTORCH_BUILD_COREML=OFF -DEXECUTORCH_BUILD_QNN=ON -DQNN_SDK_ROOT=/tmp/qnn/2.28.0.241029 -DPYTHON_EXECUTABLE=python -Bcmake-out .

cmake --build cmake-out -j9 --target install --config Release
```
rather than waiting for the full CI to run. Could try removing the `-DQNN_SDK_ROOT=/tmp/qnn/2.28.0.241029` part since it doesn't seem like it would affect this failure.

benkli01 committed:
Fix linker issue when building executor_runner

- Remove explicit addition of `-DET_EVENT_TRACER_ENABLED` from
  backends/qualcomm/CMakeLists.txt as setting the definition without
  enabling cmake flag `EXECUTORCH_ENABLE_EVENT_TRACER` caused issues
  when building the executor_runner.
- Replace deprecated namespace `torch::executor` with
  `executorch::etdump` in the executor_runner.cpp.

Signed-off-by: Benjamin Klimczak <benjamin.klimczak@arm.com>
Change-Id: Iadff38374e661f42e394dc69903548922ca08aea

benkli01 committed:
Merge branch 'pytorch:main' into add-profiling-to-xnn-executor-runner-2

benkli01 committed:
Merge branch 'pytorch:main' into add-profiling-to-xnn-executor-runner-2

dbort commented on backends/qualcomm/CMakeLists.txt:
```diff
@@ -199,11 +200,6 @@ target_link_libraries(
-#
-# add compile option
-#
-target_compile_options(executorch PUBLIC -DET_EVENT_TRACER_ENABLED)
```
@shewu-quic you added this line in https://github.com/pytorch/executorch/pull/2227/files#diff-0f6d37c62838592cf30db121c46cf34a9b9316df4b5a0d98f0ad7c7a98f7ff7eR261

It's currently causing problems because `ET_EVENT_TRACER_ENABLED` only works when `EXECUTORCH_BUILD_DEVTOOLS` is also enabled. Is it ok to remove this, or will other scripts/docs need to change?

benkli01 replied:
It should be automatically set by the main `CMakeLists.txt` as long as the CMake flag `EXECUTORCH_ENABLE_EVENT_TRACER` is enabled.

dbort commented:
@benkli01 Thank you for tracking down the problem with the Qualcomm jobs! @shewu-quic @cccclai please let us know if the change to qualcomm/CMakeLists.txt is ok.

dbort reviewed (approved):
Since the CI jobs look good, I'm fine with merging this. Thanks so much for taking the time to figure all of this out!
